### PR TITLE
Add socks5 support

### PR DESCRIPTION
**Pull request by ydylla**
Hi,
I needed a socks proxy and since I already was running caddy-l4 I thought why not add socks support for it 😄 
So this PR adds socks4 and socks5 matchers and a socks5 terminating handler.  

The socks4 matcher can match on the destination ip and port. It also matches socks4a but does not parse the domain name currently.  

The socks5 matcher is sadly very limited, since the initial data send to the server is very short. It can only match on the requested auth methods. To get to the interesting part of the handshake the matcher would have to respond and select an auth method. Theoretically this should be possible but goes far beyond initial support. Since it would also require dropping the first few response bytes from the handler (aka upstream real socks server).

Another annoyance is if you want to multiplex socks4 & socks5 like in the bellow socks.json config. The socks5 matcher has to be executed first, because if the socks4 matcher runs first it tries to read 8 bytes but a client connecting with socks5 will often only send 3 or 4 bytes causing the socks4 matcher to block indefinitely.  

Maybe it's somehow possible to add a timeout in the underling layer4 connection to better handle situations where a matcher tries to read more data than the client will initially send. Since this could also affect other matcher combinations. But currently I have no good idea.

<details>

<summary>socks.json</summary>

```json
{
  "admin": {
    "disabled": true
  },
  "logging": {
    "logs": {
      "default": {"level":"DEBUG", "encoder": {"format":"console"}}
    }
  },
  "apps": {
    "layer4": {
      "servers": {
        "socks": {
          "listen": ["0.0.0.0:1080"],
          "routes": [
            {
              "match": [
                {
                  "socks5": {}
                }
              ],
              "handle": [
                {
                  "handler": "socks5"
                }
              ]
            },
            {
              "match": [
                {
                  "socks4": {}
                }
              ],
              "handle": [
                {
                  "handler": "proxy",
                  "upstreams": [{"dial": ["10.64.0.1:1080"]}]
                }
              ]
            }
          ]
        }
      }
    }
  }
}
```

</details>

You can test this config for example with curl:
```
curl -x socks5://127.0.0.1:1080 https://ipinfo.io
```

I will also try to add tests for the handler tomorrow or next week.

### CODE REVIEW

**Comment by mholt** [comment]
Woah, this is cool! Thanks for the contribution.

Caveats noted; I'll take a look at this soon (or later, kind of hustling to get Caddy 2.6 released -- but if it takes a while to wait for feedback I may find some time)!

**Comment by mholt** [comment]
> Another annoyance is if you want to multiplex socks4 & socks5 like in the bellow socks.json config. The socks5 matcher has to be executed first, because if the socks4 matcher runs first it tries to read 8 bytes but a client connecting with socks5 will often only send 3 or 4 bytes causing the socks4 matcher to block indefinitely.

What if there was just one socks matcher? That way if both versions are configured, we have control over the ordering of the matching, and don't have to rely on the user to order them properly.

Let's see if we can combine the socks matchers.

As for not being able to get the domain name in socks5, that's probably OK for now. We can come back to that later.

Thanks for contributing this! Let me know your thoughts on combining the matchers.

**Comment by ydylla** [comment]
> What if there was just one socks matcher?

I split them on purpose to be more flexible, exactly for the case when the socks5 matcher gets extended.  
It will have to do a lot more work to fake reply and advance the handshake to the next client message. It then could also match based on username or password which the socks4 matcher will never be able to. So in the end the configuration of both will be very different. I don't see a clear way to communicate that some options only work for socks4 and some only for socks5.

Aside from that I would prefer a more general solution to the problem. If a client starts speaking socks5 to any of the following matchers ssh, xmpp, proxy_protocol, tls, socks4 the matching will block. Because all try to read more than 3 bytes via `io.ReadFull`.

Currently I experiment with prefetching and `SetReadDeadline` on the `net.Conn` before the matching phase starts. The general idea is to read as much data as the client sends or rather as much as the largest matcher as minimum needs. Then the matching only happens on buffered data and the layer4 connection could return a special error if a matcher tries to read more. Which the matching code silently interprets as no match and then continues with the next matcher.  
But right now I am not sure if this actually works.

**Comment by mholt** [comment]
While I think I understand, wouldn't it be possible to combine the matchers if the matcher first tried SOCKS4, and if the first few bytes weren't SOCKS4, then continue with the SOCKS5 matcher? It can still do its more advanced features later then. It's just a matter of trying to rule out SOCKS4 first.

Another idea: what if you call `Read()` and see how many bytes it returns? If it only returns 3-4 bytes, then try SOCKS5 maybe?

**Comment by ydylla** [comment]
Yes its possible and fairly easy to combine the matchers. Socks4 always starts with `0x04` and socks5 with `0x05` but I do not see the benefit of doing it.  
What if someone wants to accept tls/http2 and socks5 on the same port to create a hidden proxy. He also would have to put the socks5 matcher first otherwise the tls matcher will block when someone tries to connect via socks5. So it's not only sock4 vs socks5. A **not working** config for this could look like this:
<details>
<summary>hidden-socks.json</summary>

```json
{
  "apps": {
    "layer4": {
      "servers": {
        "socks": {
          "listen": ["0.0.0.0:443"],
          "routes": [
            {
              "match": [
                {
                  "tls": {}
                }
              ],
              "handle": [
                {
                  "handler": "proxy",
                  "upstreams": [{"dial": ["10.64.0.1:443"]}]
                }
              ]
            },
            {
              "match": [
                {
                  "socks5": {},
                  "ip": {"ranges": ["10.0.0.1"]}
                }
              ],
              "handle": [
                {
                  "handler": "socks5"
                }
              ]
            }
          ]
        }
      }
    }
  }
}
```

</details>
You have to switch the routes for it to work, a combined matcher that is only configured to match socks5 would also not help.  

Regarding `Read()` I already experiment with it, strangely I also get reads which return 0 bytes. I have to do some more reading/learning on how golang Reads work. Give me some days and I will likely open another PR 😄

**Comment by mholt** [comment]
Ah, I see. Thanks for explaining!

It's strange though, I feel like that config should work: if a client connects with SOCKS5, the `tls` matcher will read some bytes then return `false`, but the l4 server will "rewind" the stream so that the socks5 matcher can have a go at the same bytes, and it can read its `0x05` byte that it expects.

Does it really block forever with that config? (I haven't had a chance to try it yet.)

**Comment by ydylla** [comment]
> Does it really block forever with that config?

Yes or until the client goes away. Curl's default timeout is 5 minutes. In other words layer4 has a [Slowloris](https://en.wikipedia.org/wiki/Slowloris_(computer_security)) problem.
A total matching chain timeout would also be nice to protect against that.

**Comment by mholt** [comment]
Oh... that makes sense.

You're right, a timeout is probably a good plan for now. I'll have to think about better ways to handle this that don't involve waiting.

**Comment by mholt** [comment]
For implementing a timeout: we should probably expose a config option to set timeouts. 

Then maybe our Connection type should implement the timeout in Read(), so that any code reading from the connection will be subject to the timeout. It can just return whatever bytes it read if it times out.

**Review by mholt — COMMENTED** [review]
This is looking better. I only had time for a brief review right now. I will try to get deeper into it soon!

Still thinking about how to handle the timeout thing, or when clients don't send enough bytes...

File: README.md
Comment (ydylla) [reply README.md:186]: Why do you want to keep the tabs? In my opinion it looks terrible when the readme is rendered by GitHub where most people will probably read the examples. GitHub's tab size setting does not seam to work there for me.  
But sure it's your repo after all.

File: modules/l4socks/socks4_matcher.go
Comment (ydylla) [reply modules/l4socks/socks4_matcher.go:33]: Can you explain me why this should be changed? Is there a caddy specific reason? Because my IDE is warning me about mixed value and pointer receivers and points me to [Go FAQ](https://go.dev/doc/faq#methods_on_values_or_pointers)
> Next is consistency. If some of the methods of the type must have pointer receivers, the rest should too, so the method set is consistent regardless of how the type is used.

I will also change it for the socks5 matcher/handler which do the same.

File: README.md
Comment (mholt) [reply README.md]: Tab sizing is adjustable, spaces are not. Tabs give the power and choice to the readers.

File: modules/l4socks/socks4_matcher.go
Comment (mholt) [reply modules/l4socks/socks4_matcher.go]: :man_shrugging: I've never seen an error like that. The matcher value (m) is not ever used in the method anyway, so it shouldn't matter.

Comment (ydylla) [reply modules/l4socks/socks4_matcher.go]: It's only a weak warning 😄 So I guess it's just a matter of preference.

File: README.md
Comment (ydylla) [reply README.md]: Yes for code I agree, but apparently this is not true for code blocks in Markdown rendered by GitHub.  
But I will be strong and endure it 😭 😆

Comment (mholt) [reply README.md]: I've had a feature request open with GitHub for years. Browser extensions can do what you want though in the meantime!

**Comment by ydylla** [comment]
@mholt Is there still something blocking this? Or do you just need more time to fully review it?

**Comment by mholt** [comment]
> Is there still something blocking this?

Just the fact that I can only do one thing at a time :sweat_smile:  Sorry, I will circle back on this soon! My notifications page is full and your PRs are near the top.

File: layer4/matchers.go
Comment (mholt) [inline layer4/matchers.go:155]: Let's add a godoc comment here explaining what this does

**Review by mholt — APPROVED** [review]
Looking good to me! Just needs a godoc comment I think.

Thank you :smiley: 

I'm still thinking about that short read / timeout problem.

File: layer4/matchers.go
Comment (ydylla) [reply layer4/matchers.go:157]: Done

**Comment by ydylla** [comment]
> Just the fact that I can only do one thing at a time

No worries I understand that. Just wanted to make sure it did not forget anything.

> I'm still thinking about that short read / timeout problem

I have just added a draft PR #72 with my best attempt for that. Only the last 3 commits are relevant, I will rebase it once the others are merged.

**Comment by ydylla** [comment]
Hi @mholt,
I just fixed the conflicts. Can this be merged then?

**Comment by mholt** [comment]
Yes! Sorry

**Comment by ydylla** [comment]
Awesome, thanks

**Pull request merged**
